tools/github_update.py

```diff
@@ -96,118 +104,235 @@ def _get_repo_path(self):
+        if 'payload' in request:
+            conn.request(
+                request['method'],
+                request['path'],
+                body = json.dumps(request['payload']).encode('utf-8'),
+                headers = request_headers_with_auth)
+        else:
+            conn.request(
+                request['method'],
+                request['path'],
+                headers = request_headers_with_auth)
         return conn.getresponse()
 
 
-    def update(self, state, message, details_url = ''):
-        '''sends an update to github.
-        returns True on success or False otherwise.
-        state should be one of 'pending', 'success', 'error', or 'failure'.'''
-        logger.info('[STATUS] {} {}: {}'.format(self._context_label, state, message))
-        if details_url:
-            logger.info('[STATUS] URL: {}'.format(details_url))
+    def _pretty_duration(self, seconds):
+        """ Returns a user-friendly representation of the provided duration in seconds.
+        For example: 62.8 => "1m3s", or 129837.8 => "1d12h"
+        """
+        if seconds >= 86400:
+            # >= 1day: just do XdYh without minutes/seconds. int cast on days to avoid rounding up.
+            return '{:.0f}d{:.0f}h'.format(int(seconds / 86400), (seconds % 86400) / 3600)
+        if seconds >= 3600:
+            # >= 1hr: just do XhYm without seconds. int cast on hours to avoid rounding up.
+            return '{:.0f}h{:.0f}m'.format(int(seconds / 3600), (seconds % 3600) / 60)
 
-        if not 'WORKSPACE' in os.environ:
-            # not running in CI. skip actually sending anything to GitHub
-            return True
-        if os.environ.get('GITHUB_DISABLE', ''):
-            # environment has notifications disabled. skip actually sending anything to GitHub
-            return True
-        if not (os.environ.get('GITHUB_COMMIT_STATUS_URL') or os.environ.get('BUILD_URL')):
-            # CI job did not come from GITHUB
+        ret = ''
+        if seconds >= 60:
+            # int cast to avoid rounding up:
+            ret += '{:.0f}m'.format(int(seconds / 60))
+            seconds = seconds % 60
+        if seconds > 0:
+            # round to nearest second:
+            ret += '{:.0f}s'.format(seconds)
+        return ret
+
+
+    def _check_success(self, response, request, request_type):
+        if 200 <= response.status < 300:
             return True
+        # log failure, but don't abort the build
+        logger.error('Got {} response to {} request:'.format(response.status, request_type))
+        logger.error('Request:')
+        logger.error(pprint.pformat(request))
+        logger.error('Response:')
+        logger.error(pprint.pformat(response.read().decode('utf-8')))
+        return False
 
 
-        request = self._build_request(state, message, details_url)
+    def list_contexts(self):
+        '''returns a list of context labels of all statuses in a given commit'''
+        request = self._build_get_list_request()
```

Comment:
The following four lines should probably be abstracted into a `GithubAPI.get_status_list()` method.  It's unnecessarily low-level as it is.  Clients shouldn't have to deal with request construction and error checking.  See the `dcos` package for an example of abstracting out a remote API.